akka-cluster/src/main/scala/akka/cluster/Gossip.scala

```diff
@@ -156,59 +165,120 @@ private[cluster] final case class Gossip(
+    overview.reachability.removeObservers(membersToExclude).remove(members.collect { case m if m.team != team ⇒ m.uniqueAddress })
+  }
+
+  def teamMembers(team: Team): SortedSet[Member] =
+    members.filter(_.team == team)
+
+  def isTeamLeader(team: Team, node: UniqueAddress, selfUniqueAddress: UniqueAddress): Boolean =
+    teamLeader(team, selfUniqueAddress).contains(node)
+
+  def teamLeader(team: Team, selfUniqueAddress: UniqueAddress): Option[UniqueAddress] =
+    leaderOf(team, members, selfUniqueAddress)
 
-  def leader(selfUniqueAddress: UniqueAddress): Option[UniqueAddress] =
-    leaderOf(members, selfUniqueAddress)
+  def roleLeader(team: Team, role: String, selfUniqueAddress: UniqueAddress): Option[UniqueAddress] =
+    leaderOf(team, members.filter(_.hasRole(role)), selfUniqueAddress)
 
-  def roleLeader(role: String, selfUniqueAddress: UniqueAddress): Option[UniqueAddress] =
-    leaderOf(members.filter(_.hasRole(role)), selfUniqueAddress)
+  def leaderOf(team: Team, mbrs: immutable.SortedSet[Member], selfUniqueAddress: UniqueAddress): Option[UniqueAddress] = {
+    val reachability = teamReachability(team)
 
-  def leaderOf(mbrs: immutable.SortedSet[Member], selfUniqueAddress: UniqueAddress): Option[UniqueAddress] = {
-    val reachableMembers =
-      if (overview.reachability.isAllReachable) mbrs.filterNot(_.status == Down)
-      else mbrs.filter(m ⇒ m.status != Down &&
-        (overview.reachability.isReachable(m.uniqueAddress) || m.uniqueAddress == selfUniqueAddress))
-    if (reachableMembers.isEmpty) None
-    else reachableMembers.find(m ⇒ Gossip.leaderMemberStatus(m.status)).
-      orElse(Some(reachableMembers.min(Member.leaderStatusOrdering))).map(_.uniqueAddress)
+    val reachableTeamMembers =
+      if (reachability.isAllReachable) mbrs.filter(m ⇒ m.team == team && m.status != Down)
+      else mbrs.filter(m ⇒
+        m.team == team &&
+          m.status != Down &&
+          (reachability.isReachable(m.uniqueAddress) || m.uniqueAddress == selfUniqueAddress))
+    if (reachableTeamMembers.isEmpty) None
+    else reachableTeamMembers.find(m ⇒ Gossip.leaderMemberStatus(m.status))
+      .orElse(Some(reachableTeamMembers.min(Member.leaderStatusOrdering)))
+      .map(_.uniqueAddress)
   }
 
+  def allTeams: Set[Team] = members.map(_.team)
+
   def allRoles: Set[String] = members.flatMap(_.roles)
 
   def isSingletonCluster: Boolean = members.size == 1
 
+  /**
+   * @return true if fromAddress should be able to reach toAddress based on the unreachability data and their
+   *         respective teams
+   */
+  def isReachableExcludingDownedObservers(fromTeam: Team, toAddress: UniqueAddress): Boolean =
+    if (!hasMember(toAddress)) false
+    else {
+      val to = member(toAddress)
+
+      // if member is in the same team, we ignore cross-team unreachability
+      if (fromTeam == to.team) teamReachabilityExcludingDownedObservers(fromTeam).isReachable(toAddress, toAddress)
```

Comment:
It looks like a mistake, so I guessed you are right, but if I change it the cross team unreachability assertions fails in gossip spec. Hmmm.